external/fv3fit/fv3fit/keras/_models/dense.py

```diff
@@ -0,0 +1,235 @@
+    # ensures [sample, z] dimensionality
+    if len(array.shape) == 2:
+        return array
+    elif len(array.shape) == 1:
+        return array[:, None]
+    else:
+        raise ValueError(f"expected 1d or 2d array, got shape {array.shape}")
+
+
+def _get_input_layer_shapes(X: Sequence[np.ndarray]) -> List[Tuple[int]]:
+    # adds a z dim of length 1 for non vertical features, so that they can
+    # be concantenated with inputs that have a z dimension
+    shapes: List[Tuple[int]] = []
+    for array in X:
+        array_features = _count_array_features(array)
+        sample_shape: Tuple[int] = array.shape[1:]
+        if array_features == 1:
+            shapes.append((*sample_shape, 1))  # type: ignore
+        else:
+            shapes.append(sample_shape)
+    return shapes
+
+
+def build_model(
+    config: DenseHyperparameters, X, y
+) -> Tuple[tf.keras.Model, tf.keras.Model]:
+    """
+    Args:
+        config: configuration of convolutional training
+        X: example input for keras fitting, used to determine shape and normalization
+        y: example output for keras fitting, used to determine shape and normalization
+    """
+    input_layer_shapes = _get_input_layer_shapes(X)
+    input_layers = [
+        tf.keras.layers.Input(shape=input_shape) for input_shape in input_layer_shapes
+    ]
+    norm_input_layers = standard_normalize(
+        names=config.input_variables,
+        layers=input_layers,
+        arrays=[_ensure_2d(array) for array in X],
+    )
+    if len(norm_input_layers) > 1:
+        full_input = tf.keras.layers.Concatenate()(norm_input_layers)
+    else:
+        full_input = norm_input_layers[0]
+
+    hidden_outputs = []
+
+    x = full_input
+    for i in range(config.dense_network.depth - 1):
+        if config.dense_network.gaussian_noise > 0.0:
+            x = tf.keras.layers.GaussianNoise(
+                config.dense_network.gaussian_noise, name=f"gaussian_noise_{i}"
+            )(x)
+        hidden_layer = tf.keras.layers.Dense(
+            config.dense_network.width,
+            activation=tf.keras.activations.relu,
+            kernel_regularizer=config.dense_network.kernel_regularizer.instance,
+            name=f"hidden_{i}",
+        )
+        if config.dense_network.spectral_normalization:
+            hidden_layer = tfa.layers.SpectralNormalization(
+                hidden_layer, name=f"spectral_norm_{i}"
+            )
+        x = hidden_layer(x)
+        hidden_outputs.append(x)
+
+    norm_output_layers = [
+        tf.keras.layers.Dense(
+            _count_array_features(array),
+            activation="linear",
+            name=f"dense_network_output_{i}",
+        )(hidden_outputs[-1])
+        for i, array in enumerate(y)
+    ]
+    if config.nonnegative_outputs is True:
+        norm_output_layers = [
+            tf.keras.layers.Activation(tf.keras.activations.relu)(output_layer)
+            for output_layer in norm_output_layers
+        ]
```

Comment:
This is a bug - you need to apply this activation to `denorm_output_layers`. Non-negativity in normalized space doesn't guarantee non-negativity in denormalized space.

Reply:
Thanks, good catch